docs/source/cpp/streaming_execution.rst

```diff
@@ -175,9 +175,607 @@ their completion::
+
+
+Creating a Scan `ExecNode`::
+
+    auto options = std::make_shared<arrow::dataset::ScanOptions>();
+    options->use_async = true; 
+    options->projection = Materialize({});  // create empty projection
+
+    // construct the scan node
+    cp::ExecNode* scan;
+    auto scan_node_options = arrow::dataset::ScanNodeOptions{dataset, options};
+
+    ARROW_ASSIGN_OR_RAISE(scan,
+                            cp::MakeExecNode("scan", plan.get(), {}, 
+                              scan_node_options));
+
+
+Write-Node
+----------
+
+The option to write a result to a file format is provided by this execution node type. 
+A definition doesn't exist as an :class:`ExecNode`, but the write options are provided
+via the :class:`arrow::dataset::WriteNodeOptions` and defined using 
+:class::`arrow::dataset::FileSystemDatasetWriteOptions`, `std::shared_ptr<arrow::Schema>`,
+and `std::shared_ptr<arrow::util::AsyncToggle> backpressure_toggle`. Here the 
+:class::`arrow::dataset::FileSystemDatasetWriteOptions` contains the meta-data required 
+to write the data. 
+
+Creating `WriteNodeOptions`::
+
+    std::string root_path = "";
+    std::string uri = "file://" + '/path/to/file';
+    std::shared_ptr<arrow::fs::FileSystem> filesystem =
+    arrow::fs::FileSystemFromUri(uri, &root_path).ValueOrDie();
+
+    auto base_path = root_path + "/parquet_dataset";
+    ABORT_ON_FAILURE(filesystem->DeleteDir(base_path));
+    ABORT_ON_FAILURE(filesystem->CreateDir(base_path));
+
+    // The partition schema determines which fields are part of the partitioning.
+    auto partition_schema = arrow::schema({arrow::field("a", arrow::int32())});
+    // We'll use Hive-style partitioning,
+    // which creates directories with "key=value" pairs.
+
+    auto partitioning =
+    std::make_shared<arrow::dataset::HivePartitioning>(partition_schema);
+    // We'll write Parquet files.
+    auto format = std::make_shared<arrow::dataset::ParquetFileFormat>();
+
+    arrow::dataset::FileSystemDatasetWriteOptions write_options;
+    write_options.file_write_options = format->DefaultWriteOptions();
+    write_options.filesystem = filesystem;
+    write_options.base_dir = base_path;
+    write_options.partitioning = partitioning;
+    write_options.basename_template = "part{i}.parquet";
+
+    arrow::dataset::WriteNodeOptions write_node_options {write_options,
+    dataset->schema()};
+
+Creating a `write` `ExecNode`::
+
+    ARROW_ASSIGN_OR_RAISE(cp::ExecNode *wr, cp::MakeExecNode("write", plan.get(),
+        {scan}, write_node_options));
+
+    ABORT_ON_FAILURE(wr->Validate());
+    ABORT_ON_FAILURE(plan->Validate());
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+    plan->finished().Wait(); // make sure to add this method 
+
+``UnionNode``
+-------------
+
+:class:`UnionNode` is the :class:`ExecNode` interface to perform a union 
+operation on two datasets. The union operation can be executed
```

Comment:
This can be confusing given the various uses of union in SQL. Specifically, UnionNode combines two streams of data with the same schema into one, and does not modify or look at any of the data itself.